File: test/fixtures/sample.rs
// Sample Rust file for LSP testing with rust-analyzer

use std::collections::HashMap;
use std::io::{self, Write};
use std::sync::{Arc, Mutex};

// Struct with derive macros
#[derive(Debug, Clone, PartialEq)]
pub struct Person {
    name: String,
    age: u32,
    email: Option<String>,
}

impl Person {
    // Constructor
    pub fn new(name: impl Into<String>, age: u32) -> Self {
        Self {
            name: name.into(),
            age,
            email: None,
        }
    }
    
    // Method with mutable self
    pub fn set_email(&mut self, email: String) {
        self.email = Some(email);
    }
    
    // Method with immutable self
    pub fn can_vote(&self) -> bool {
        self.age >= 18
    }
    
    // Intentional error: undefined field
    pub fn get_phone(&self) -> &str {
        &self.phone  // Error: no field `phone`
    }
}

// Generic struct
struct Container<T> {
    items: Vec<T>,
}

impl<T> Container<T> {
    fn new() -> Self {
        Container { items: Vec::new() }
    }
    
    fn add(&mut self, item: T) {
        self.items.push(item);
    }
    
    // Test completion here: self.items.
}

// Trait definition
trait Drawable {
    fn draw(&self);
    
    // Default implementation
    fn description(&self) -> String {
        String::from("A drawable object")
    }
}

// Enum with data
#[derive(Debug)]
enum Message {
    Text(String),
    Image { url: String, width: u32, height: u32 },
    Video(String, u32),  // url, duration
}

impl Message {
    fn process(&self) {
        match self {
            Message::Text(content) => println!("Text: {}", content),
            Message::Image { url, width, height } => {
                println!("Image: {} ({}x{})", url, width, height);
            }
            Message::Video(url, duration) => {
                println!("Video: {} ({} seconds)", url, duration);
            }
            // Missing pattern (potential warning)
        }
    }
}

// Async function
async fn fetch_data(url: &str) -> Result<String, Box<dyn std::error::Error>> {
    // Simulated async operation
    tokio::time::sleep(std::time::Duration::from_millis(100)).await;
    
    // Type error: wrong return type
    Ok(42)  // Should return String
}

// Function with lifetime parameters
fn longest<'a>(x: &'a str, y: &'a str) -> &'a str {
    if x.len() > y.len() {
        x
    } else {
        y
    }
}

// Closure examples
fn closure_examples() {
    let numbers = vec![1, 2, 3, 4, 5];
    
    // Simple closure
    let sum: i32 = numbers.iter().sum();
    
    // Closure with type annotation
    let is_even = |x: &i32| -> bool { x % 2 == 0 };
    
    // Filter and map
    let even_squares: Vec<i32> = numbers
        .iter()
        .filter(|&&x| is_even(&x))
        .map(|&x| x * x)
        .collect();
    
    // Move closure
    let captured = String::from("Hello");
    let print_captured = move || {
        println!("{}", captured);
    };
    
    // Error: captured moved
    println!("{}", captured);  // Error: value moved
}

// Macro definition
macro_rules! create_function {
    ($func_name:ident) => {
        fn $func_name() {
            println!("Called function: {}", stringify!($func_name));
        }
    };
}

create_function!(foo);
create_function!(bar);

// Error handling with ? operator
fn read_config() -> Result<Config, ConfigError> {
    let content = std::fs::read_to_string("config.toml")?;
    let config: Config = toml::from_str(&content)?;
    Ok(config)
}

// Undefined types (errors)
struct Config;
struct ConfigError;

// Formatting test - poorly formatted
fn   poorly_formatted(  x:i32,y:i32  )->i32{x+y}

// Main function
fn main() {
    let mut person = Person::new("Alice", 25);
    person.set_email("alice@example.com".to_string());
    
    // Pattern matching
    let msg = Message::Text("Hello, Rust!".to_string());
    msg.process();
    
    // Error handling
    match read_config() {
        Ok(config) => println!("Config loaded"),
        Err(e) => eprintln!("Error: {}", e),
    }
    
    // Thread safety with Arc and Mutex
    let counter = Arc::new(Mutex::new(0));
    let counter_clone = Arc::clone(&counter);
    
    std::thread::spawn(move || {
        let mut num = counter_clone.lock().unwrap();
        *num += 1;
    });
    
    // Missing semicolon
    println!("Counter: {:?}", counter.lock().unwrap())
    
    // Unreachable code
    return;
    println!("This is unreachable");
}

File: test/fixtures/sample.cpp
// Sample C++ file for LSP testing with clangd
#include <algorithm>
#include <iostream>
#include <map>
#include <memory>
#include <string>
#include <vector>

// Template class for testing
template <typename T>
class Container {
 private:
  std::vector<T> items;
  mutable size_t access_count = 0;

 public:
  Container() = default;

  // Add item to container
  void add(const T& item) { items.push_back(item); }

  // Add item with move semantics
  void add(T&& item) { items.push_back(std::move(item)); }

  // Get item at index
  const T& get(size_t index) const {
    ++access_count;
    return items.at(index);  // May throw std::out_of_range
  }

  // Get container size
  [[nodiscard]] size_t size() const noexcept { return items.size(); }

  // Iterator support
  auto begin() { return items.begin(); }
  auto end() { return items.end(); }
  auto begin() const { return items.cbegin(); }
  auto end() const { return items.cend(); }

  // Intentional error: undefined member function
  void undefined_method() {
    undefined_variable = 42;  // Error: undefined variable
  }
};

// Class with inheritance
class Shape {
 public:
  virtual ~Shape() = default;
  virtual double area() const = 0;
  virtual void draw() const = 0;
};

class Rectangle : public Shape {
 private:
  double width_, height_;

 public:
  Rectangle(double w, double h) : width_(w), height_(h) {}

  double area() const override { return width_ * height_; }

  void draw() const override {
    std::cout << "Drawing rectangle: " << width_ << "x" << height_ << "\n";
  }

  // Missing override keyword (potential warning)
  virtual double perimeter() const { return 2 * (width_ + height_); }
};

// Function with auto return type
auto process_data(const std::vector<int>& data) {
  std::map<std::string, int> result;

  // Test completion here: std::

  for (const auto& value : data) {
    if (value % 2 == 0) {
      result["even"] += 1;
    } else {
      result["odd"] += 1;
    }
  }

  // Type error: returning wrong type
  return data;  // Should return result
}

// Template function
template <typename InputIt, typename UnaryPredicate>
auto count_matching(InputIt first, InputIt last, UnaryPredicate pred) {
  return std::count_if(first, last, pred);
}

// Formatting test - poorly formatted function
void poorly_formatted(int x, int y, int z) {
  int result = x + y + z;
  return;
}

// Modern C++ features
class ModernClass {
  std::unique_ptr<int[]> data;
  size_t size_;

 public:
  // Constructor with initializer list
  ModernClass(std::initializer_list<int> init)
      : data(std::make_unique<int[]>(init.size())), size_(init.size()) {
    std::copy(init.begin(), init.end(), data.get());
  }

  // Rule of 5
  ModernClass(const ModernClass&) = delete;
  ModernClass& operator=(const ModernClass&) = delete;
  ModernClass(ModernClass&&) = default;
  ModernClass& operator=(ModernClass&&) = default;
  ~ModernClass() = default;

  // Range-based for loop support
  int* begin() { return data.get(); }
  int* end() { return data.get() + size_; }
};

// Constexpr function
constexpr int factorial(int n) { return (n <= 1) ? 1 : n * factorial(n - 1); }

// Main function
int main() {
  // Test container
  Container<std::string> strings;
  strings.add("Hello");
  strings.add("World");

  // Test shape hierarchy
  auto rect = std::make_unique<Rectangle>(10.0, 5.0);
  std::cout << "Area: " << rect->area() << std::endl;

  // Test modern features
  ModernClass modern{1, 2, 3, 4, 5};

  // Lambda expression
  auto is_positive = [](int x) { return x > 0; };

  std::vector<int> numbers{-2, -1, 0, 1, 2, 3};
  auto positive_count = count_matching(numbers.begin(), numbers.end(), is_positive);

  // Structured binding (C++17)
  auto [key, value] = *result.begin();  // Error: result not in scope

  // Missing semicolon
  std::cout << "Positive numbers: " << positive_count
            << "\n"

      return 0;
}
